### PR TITLE
[Pytorch Edge][tracing-based] reorganize model tracer dependency

### PR DESCRIPTION
**Pull request by cccclai**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66237
* #64152
* #64087
* __->__ #63421

1. move 4 files to :
```
KernelDTypeTracer.h
KernelDTypeTracer.h
OperatorCallTracer.h
OperatorCallTracer.h
```
so it's visible in OSS.

2. Update the namespace to `torch::jit::mobile`
3. Add a `fb_xplat_cxx_library` `torch_model_tracer` with the source file list above.
4. update the `fb_xplat_cxx_library`  `model_tracer_lib` dependency on the new `torch_model_tracer` library

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30378069/)!

Differential Revision: [D30378069](https://our.internmc.facebook.com/intern/diff/D30378069)

### CODE REVIEW

File: torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * You can get the set of operators that were called (op_name.overload_name)
+ * using getCalledOperators().
+ *
+ * Note: This class is not thread safe or re-entrant, and should not be used
+ * across multiple threads of execution.
+ *
+ */
+struct OperatorCallTracer final {
+  static std::set<std::string> called_operators_;
```
Comment (malfet) [inline torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.h:20]: Does this set really needs to be ordered? I.e. wouldn't `std::unordered_set` be faster?

File: torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  at::CallbackHandle handle_;
+
+  OperatorCallTracer() {
```
Comment (malfet) [inline torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.h:23]: Is `OperatorCallTracer` supposed to be a singleton? If so, wouldn't it be better to make its constructor private?

File: torch/csrc/jit/mobile/model_tracer/KernelDTypeTracer.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ */
+struct KernelDTypeTracer final {
+  at::CallbackHandle handle_;
+  /* The key of the map below (std::string) is the kernel tag name (constant
+   * character string) which shows up in code. The value part of type
+   * std::set<std::string> is the collection of dtypes for which we need to
+   * generate code for the said kernel tag.
+   */
+  typedef std::map<std::string, std::set<std::string>> kernel_tags_type;
```
Comment (malfet) [inline torch/csrc/jit/mobile/model_tracer/KernelDTypeTracer.h:29]: Is order of keys in a map important? If not, please use unordered map and unordered set
```suggestion
  typedef std::unordered_map<std::string, std::unordered_set<std::string>> kernel_tags_type;
```

File: torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.h (hunk 1 above)
Comment (JacobSzwejbka) [reply torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.h:20]: Its expected for these traced operators to be serialized so its convenient for the ordering to always be consistent. The efficiency difference should be small, the majority of models in my experience have ~100 operators.

**Comment by cccclai** [comment]
Thanks Nikita for the feedback. Let me have a separate pr on top of the pr stack to address them.

**Pull request closed without merging**